### PR TITLE
EventGrid - Add api-version 2020-10-15-preview 

### PR DESCRIPTION
**Pull request by batrived**
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Please ensure to add changelog with this PR by answering the following questions.
  1. What's the purpose of the update?    
      - [ ] new service onboarding 
      - [x] new API version 
      - [ ] update existing version for new feature 
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify 
  2. When you are targeting to deploy new service/feature to public regions? Please provide date, or month to public if date is not available yet.
  3. When you expect to publish swagger? Please provide date, or month to public if date is not available yet.
  4. If it's an update to existing version,  please select SDKs of specific language and CLIs that require refresh after swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No, no need to refresh for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of “no breaking changes
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [x] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “WaitForARMFeedback” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  - Adding a new service

- [x] Please ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [x] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from API Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable version with Breaking Change Validation errors
- [ ] Updating API(s) in preview over 1 year

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://msazure.visualstudio.com/One/_workitems/create/Scenario?templateId=d1c215ec-5d6a-4cb3-a193-95ddd0a1ba96&ownerId=85bb92fb-21fe-41ca-9309-47f2c5f57515). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

**Comment by batrived** [comment]
LINT errors are for existing swagger definitions and I haven't made changes in that area. We have a workitem to fix the existing swagger changes, so we need exemption on this PR

File: specification/eventgrid/resource-manager/Microsoft.EventGrid/preview/2020-10-15-preview/EventGrid.json
Comment (j5lim) [inline specification/eventgrid/resource-manager/Microsoft.EventGrid/preview/2020-10-15-preview/EventGrid.json:5665]: ARM requires a new top level property `systemData` for all tracked / proxy resources.
https://armwiki.azurewebsites.net/api_contracts/guidelines/rpc.html#rpc027-systemdata-support

You can reference the one in common-type.
https://github.com/Azure/azure-rest-api-specs/blob/a16488bbd0d17551c729e22ea6ac6f7c6588a0a3/specification/common-types/resource-management/v2/types.json#L445

File: specification/eventgrid/resource-manager/Microsoft.EventGrid/preview/2020-10-15-preview/EventGrid.json
Comment (j5lim) [inline specification/eventgrid/resource-manager/Microsoft.EventGrid/preview/2020-10-15-preview/EventGrid.json:6687]: Where is this referenced?

File: specification/eventgrid/resource-manager/Microsoft.EventGrid/preview/2020-10-15-preview/EventGrid.json
Comment (j5lim) [inline specification/eventgrid/resource-manager/Microsoft.EventGrid/preview/2020-10-15-preview/EventGrid.json:7]: Please define schema for default error response.

**Review by j5lim — COMMENTED** [review]
> LINT errors are for existing swagger definitions and I haven't made changes in that area. We have a workitem to fix the existing swagger changes, so we need exemption on this PR

Can the errors be fixed now? It will be easier than fixing after the api-version is published. If it is time-constrained, can you please provide the work item and ETA for fixing it? Thanks.

File: specification/eventgrid/resource-manager/Microsoft.EventGrid/preview/2020-10-15-preview/EventGrid.json
Comment (batrived) [reply specification/eventgrid/resource-manager/Microsoft.EventGrid/preview/2020-10-15-preview/EventGrid.json:5665]: @j5lim  This seems like a new requirement and requires additional work and testing. It is outside the scope of this change as I am not even adding a new resource type. 
I will create a work item for us to investigate this new property and add that as a separate PR.

File: specification/eventgrid/resource-manager/Microsoft.EventGrid/preview/2020-10-15-preview/EventGrid.json
Comment (batrived) [reply specification/eventgrid/resource-manager/Microsoft.EventGrid/preview/2020-10-15-preview/EventGrid.json:6687]: @j5lim This is a derived class of "DeliveryAttributeMapping" which is being referenced in "EventHubEventSubscriptionDestinationProperties" (line 6824)

**Comment by batrived** [comment]
> > LINT errors are for existing swagger definitions and I haven't made changes in that area. We have a workitem to fix the existing swagger changes, so we need exemption on this PR
> 
> Can the errors be fixed now? It will be easier than fixing after the api-version is published. If it is time-constrained, can you please provide the work item and ETA for fixing it? Thanks.

This is the workitem: https://msazure.visualstudio.com/Swagger%20KPIs/_workitems/edit/8187847 and it is marked to be done in December. 
The reason I am not fixing in this PR is because we have an existing workitem and that will need some testing from our side which is out of the scope for these changes

**Comment by j5lim** [comment]
> > > LINT errors are for existing swagger definitions and I haven't made changes in that area. We have a workitem to fix the existing swagger changes, so we need exemption on this PR
> > 
> > 
> > Can the errors be fixed now? It will be easier than fixing after the api-version is published. If it is time-constrained, can you please provide the work item and ETA for fixing it? Thanks.
> 
> This is the workitem: https://msazure.visualstudio.com/Swagger%20KPIs/_workitems/edit/8187847 and it is marked to be done in December.
> The reason I am not fixing in this PR is because we have an existing workitem and that will need some testing from our side which is out of the scope for these changes

Thanks. Submitting separate PRs is fine. I just wanted to point that fixing those later may require additional process if it is considered as a breaking change.

**Review by j5lim — APPROVED** [review]
Signed off from ARM side.

File: specification/eventgrid/resource-manager/readme.md
Comment (akning-ms) [inline specification/eventgrid/resource-manager/readme.md:29]: Will this version be generated for SDK? if yes. default tag should change to package-2020-10-preview, so we can have correct sdk validation during PR review phase

File: specification/eventgrid/resource-manager/Microsoft.EventGrid/preview/2020-10-15-preview/EventGrid.json
Comment (akning-ms) [inline specification/eventgrid/resource-manager/Microsoft.EventGrid/preview/2020-10-15-preview/EventGrid.json:2668]: can we ref to default errorReponse in common type v2(ARM standard definition)? https://github.com/Azure/azure-rest-api-specs/blob/1c100daabed6a008c76f3e4e0cae4e7eac5a5bf4/specification/common-types/resource-management/v2/types.json#L309

File: specification/eventgrid/resource-manager/readme.md
Comment (batrived) [reply specification/eventgrid/resource-manager/readme.md]: Yes we will have SDK for this new api version. I have made the changes.

File: specification/eventgrid/resource-manager/Microsoft.EventGrid/preview/2020-10-15-preview/EventGrid.json
Comment (batrived) [reply specification/eventgrid/resource-manager/Microsoft.EventGrid/preview/2020-10-15-preview/EventGrid.json:2668]: thanks for the suggestion. We will keep this as is as these are auto generated for us using swashbuckel and we would like to keep the same pattern with other existing resources..

**Comment by akning-ms** [comment]
> > > LINT errors are for existing swagger definitions and I haven't made changes in that area. We have a workitem to fix the existing swagger changes, so we need exemption on this PR
> > 
> > 
> > Can the errors be fixed now? It will be easier than fixing after the api-version is published. If it is time-constrained, can you please provide the work item and ETA for fixing it? Thanks.
> 
> This is the workitem: https://msazure.visualstudio.com/Swagger%20KPIs/_workitems/edit/8187847 and it is marked to be done in December.
> The reason I am not fixing in this PR is because we have an existing workitem and that will need some testing from our side which is out of the scope for these changes

Hi @batrived, can we know what is your target version/time to fix it, actually  it is good chance to fix it in this PR this new api version.  if this PR merged, it mean you need a new API version to fix above lint or correctness issue in above mentioned workitem. and breaking change review board just opened a window to fix existing S360 Swagger KPI, it means you can get approve faster. :)

**Comment by batrived** [comment]
> > > > LINT errors are for existing swagger definitions and I haven't made changes in that area. We have a workitem to fix the existing swagger changes, so we need exemption on this PR
> > > 
> > > 
> > > Can the errors be fixed now? It will be easier than fixing after the api-version is published. If it is time-constrained, can you please provide the work item and ETA for fixing it? Thanks.
> > 
> > 
> > This is the workitem: https://msazure.visualstudio.com/Swagger%20KPIs/_workitems/edit/8187847 and it is marked to be done in December.
> > The reason I am not fixing in this PR is because we have an existing workitem and that will need some testing from our side which is out of the scope for these changes
> 
> Hi @batrived, can we know what is your target version/time to fix it, actually it is good chance to fix it in this PR this new api version. if this PR merged, it mean you need a new API version to fix above lint or correctness issue in above mentioned workitem. and breaking change review board just opened a window to fix existing S360 Swagger KPI, it means you can get approve faster. :)

Thank you for your suggestion and concern. We would like to follow the proper way to fix this and my teammate already working on this. We would like to separate this review and fixes to include the items/works our customer needs without changing or impacting other areas. Appreciate the feedback

**Review by ahamad-MS — APPROVED** [review]

**Comment by ahamad-MS** [comment]
@akning-ms Can you please provide an ETA for your approval? @batrived already provided all the required info needed and we need to merge this PR ASAP to start the next wave of SDK work. Your help in appreciated.

thanks

File: specification/eventgrid/resource-manager/Microsoft.EventGrid/preview/2020-10-15-preview/EventGrid.json
Comment (lirenhe) [inline specification/eventgrid/resource-manager/Microsoft.EventGrid/preview/2020-10-15-preview/EventGrid.json:27]: I saw the global definition of "consumes" and 'produces',  do you still need to add it into each operation?

File: specification/eventgrid/resource-manager/Microsoft.EventGrid/preview/2020-10-15-preview/EventGrid.json
Comment (lirenhe) [inline specification/eventgrid/resource-manager/Microsoft.EventGrid/preview/2020-10-15-preview/EventGrid.json:130]: do you need to define the following?
        "x-ms-long-running-operation-options": {
          "final-state-via": "location"
        }

File: specification/eventgrid/resource-manager/Microsoft.EventGrid/preview/2020-10-15-preview/EventGrid.json
Comment (lirenhe) [inline specification/eventgrid/resource-manager/Microsoft.EventGrid/preview/2020-10-15-preview/EventGrid.json:61]: I saw in your previous swagger, the default response is defined as below.  Could you make the behavior consistent?
          "default": {
            "description": "Eventhub error response describing why the operation failed.",
            "schema": {
              "$ref": "../../../common/v1/definitions.json#/definitions/ErrorResponse"
            }
          }

**Review by majastrz — APPROVED** [review]
The delta since last review (making regenerateKeys async) looks good. Still signed off from ARM side.

File: specification/eventgrid/resource-manager/Microsoft.EventGrid/preview/2020-10-15-preview/EventGrid.json
Comment (ahamad-MS) [reply specification/eventgrid/resource-manager/Microsoft.EventGrid/preview/2020-10-15-preview/EventGrid.json:130]: We are not familiar with this property and/or what it is used for. Please note that domain is an existing resource and this definition is not new (even though it shows as new in this iteration). We will not change existing behavior/definition as this is PR as this is outside the scope of the change and we need to ensure we understand the behavior correctly to apply to all resources.. Hence, Won't fix here.

File: specification/eventgrid/resource-manager/Microsoft.EventGrid/preview/2020-10-15-preview/EventGrid.json
Comment (ahamad-MS) [reply specification/eventgrid/resource-manager/Microsoft.EventGrid/preview/2020-10-15-preview/EventGrid.json:27]: It is not clear to us what is the ask here. Please note that this is auto-generated code/spec for us and this is existing resource/section from previous swagger (please refer to https://github.com/Azure/azure-rest-api-specs/blob/master/specification/eventgrid/resource-manager/Microsoft.EventGrid/preview/2020-04-01-preview/EventGrid.json#L12 )... 

This is outside scope of this PR. We will not fix it as part of this PR as it is specific for new features/resources. If any fix, it should be through the provided tools for services team. thanks

File: specification/eventgrid/resource-manager/Microsoft.EventGrid/preview/2020-10-15-preview/EventGrid.json
Comment (ahamad-MS) [reply specification/eventgrid/resource-manager/Microsoft.EventGrid/preview/2020-10-15-preview/EventGrid.json:61]: Same as above. This is not new resource/change. This already exists as part of previous swagger and not part of this PR.

PLease refere to: https://github.com/Azure/azure-rest-api-specs/blob/master/specification/eventgrid/resource-manager/Microsoft.EventGrid/preview/2020-04-01-preview/EventGrid.json#L60

Also, this is auto-generated for us and it is consistent in all other resources in our sawagger. Hence, Won't fix. thanks.

**Comment by ahamad-MS** [comment]
@lirenhe  all your comments are addressed and they are outside scope of this PR as they are about existing resources in previous swagger. We would need your approval along with other reviewers today to allow us to meet our deadlines.

If additional questions needed, please let us discuss quickly in person today. Otherwise, kindly approve and merge the changes if no more issues for new resources.

thanks
Ashraf

**Review by Kishp01 — APPROVED** [review]

**Comment by akning-ms** [comment]
> @akning-ms Can you please provide an ETA for your approval? @batrived already provided all the required info needed and we need to merge this PR ASAP to start the next wave of SDK work. Your help in appreciated.
> 
> thanks
Thanks, understand your case,  for  R4009 - RequiredSystemDataInNewApiVersions we leave it for you to implement in next version.
But for below:

❌ R4010 - RequiredDefaultResponse | The response is defined but without a default error response implementation.Consider adding it.'New: Microsoft.EventGrid/preview/2020-10-15-preview/EventGrid.json#L5345
-- | --
❌ R4011 - DeleteOperationResponses | The delete operation is defined without a 200 or 204 error response implementation,please add it.'New: Microsoft.EventGrid/preview/2020-10-15-preview/EventGrid.json#L1384
❌ R4011 - DeleteOperationResponses | The delete operation is defined without a 200 or 204 error response implementation,please add it.'New: Microsoft.EventGrid/preview/2020-10-15-preview/EventGrid.json#L1818
❌ R4011 - DeleteOperationResponses | The delete operation is defined without a 200 or 204 error response implementation,please add it.'New: Microsoft.EventGrid/preview/2020-10-15-preview/EventGrid.json#L2981
❌ R4011 - DeleteOperationResponses | The delete operation is defined without a 200 or 204 error response implementation,please add it.'New: Microsoft.EventGrid/preview/2020-10-15-preview/EventGrid.json#L4614


Suggest you fix it first as it may not too difficult to fix which related with correct response. because once you merge this PR into master, we take it as release. then especially you release this error in SDK. then it will be difficult to fix as it become a breaking change to customer.

**Comment by akning-ms** [comment]
Ping @batrived for address Lint R4011 issue

**Comment by batrived** [comment]
> EventGrid.jso

@akning-ms I don't know why these Lint errors are being thrown because I checked those apis and they do return "204" and Lint error is saying that I need 202 OR 204. Could you please help me understand these errors?

**Comment by akning-ms** [comment]
> > EventGrid.jso
> 
> @akning-ms I don't know why these Lint errors are being thrown because I checked those apis and they do return "204" and Lint error is saying that I need 202 OR 204. Could you please help me understand these errors?
Hi @batrived , you can refer to https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/openapi-authoring-automated-guidelines.md#R4011 for how to fix. basically, you missed 200 response code.

More detail for all linter issue, you can refer to:
https://github.com/Azure/azure-rest-api-specs/pull/11879/checks?check_run_id=1731937635

**Comment by ahamad-MS** [comment]
Hello @akning-ms  all linter issues are addressed now.. Can you please help merge this PR .. thanks

**Review by ahamad-MS — APPROVED** [review]
LGTM

**Pull request merged**
